### PR TITLE
Flow triggers 1/2: no-wait cases

### PR DESCRIPTION
**Pull request by hjoliver**
These changes partially address #4686 

- implement agreed final `cylc trigger` behaviour, **sans `--wait` option** (see #4739 for that)
- extend reference testing to flow numbers (w/ back-compat for old logs)
- add tests based on @oliver-sanders examples in #4686 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.

- [x] Appropriate tests are included (unit and/or functional).

- [x] Appropriate change log entry included.

- [x] (master branch) I have opened a documentation PR at https://github.com/cylc/cylc-doc/pull/427

### CODE REVIEW

**Comment by hjoliver** [comment]
Now that we don't require `--wait` to prevent reflow (*of active flows*) this could be a standalone PR, with `--wait` (and its DB implications etc.) as a follow-up.  (And maybe get this into rc2 if it's all good).

**Comment by hjoliver** [comment]
(No doubt I have broken some existing tests, with changes to `cylc trigger` and associated API args. Should be easy fixes though).

**Comment by hjoliver** [comment]
@oliver-sanders or @dpmatthews - plz assign another reviewer at your end.

**Comment by hjoliver** [comment]
Fast tests will pass after #4746

**Comment by oliver-sanders** [comment]
(FYI: will review soon, just getting the RC2 bugs out the way first)

File: cylc/flow/scripts/trigger.py [hunk 1]
```diff
@@ -96,20 +97,60 @@ def get_option_parser():
...
+        help="description of triggered flow (with --flow=new) ."
     )
 
     return parser
 
 
+def check_flow_options(opt_flow, opt_flow_descr):
+    """Check validity of flow-related options.
+
+    Examples:
```
Comment (hjoliver) [inline cylc/flow/scripts/trigger.py:119]: NOTE in 2/2 I convert these doctests to separate unit tests.

Comment (hjoliver) [inline cylc/flow/scripts/trigger.py:119]: NOTE in 2/2 I convert these doctests to separate unit tests.

**Comment by dwsutherland** [comment]
Had a play with this one, using this workflow:
```
[cylc]
    cycle point format = CCYYMMDDThh
    UTC mode = True

[scheduling]
    initial cycle point = 20210101T00
    [[special tasks]]
        clock-trigger = foo(PT0H)
    [[graph]]
        P1M = """
foo => bar => qux => qaz
foo[-P1M] => foo
"""
```


Triggering in the future ahead of the first flow worked as expected:
`$ cylc trigger --flow=all linear //20211001T00/foo`
as was default triggering ahead.
Triggering behind the first flow worked as expected:
`$ cylc trigger linear //20210301T00/foo`
```
          {
            "id": "~sutherlander/linear/run1//20210301T00/foo",
            "flowNums": "[1]",
            "jobSubmits": 2,
            "state": "running",
            "jobs": [
              {
                "id": "~sutherlander/linear/run1//20210301T00/foo/01"
              },
              {
                "id": "~sutherlander/linear/run1//20210301T00/foo/02"
              }
            ],
            "prerequisites": [
              {
                "satisfied": false,
                "expression": "c0",
                "conditions": [
                  {
                    "exprAlias": "c0",
                    "taskId": "20210201T00/foo",
                    "satisfied": false
                  }
                ]
              }
            ]
          },
```
Reflow worked as expected;
`$ cylc trigger --flow=new linear //20210301T00/foo`
however the new flow stalled where I `--flow=all` triggered between the reflow and original flow front:
`$ cylc trigger --flow=all linear //20211001T00/foo`

Is this because the reflow get's absorbed by the `[1, 2]` trigger?

**Comment by hjoliver** [comment]
> Is this because the reflow get's absorbed by the [1, 2] trigger?

Yes, if you triggered with `--flow=all` (i.e. the default) in front of the new flow, then you've started another front for all existing flows including the new one. (So, the new flow will consider that tasks in the new front have already run, when it reaches that part of the graph).

Worth a look at the cylc-doc PR (referenced above) if you haven't already.

File: cylc/flow/scripts/trigger.py [hunk 2]
```diff
@@ -27,30 +27,25 @@
...
 
-Other tasks (those outside of the n=0 graph window) do not yet belong to a
-flow. Manual triggering of these with --reflow will start a new flow; otherwise
-only the triggered task will run.
+If a task is in the n=0 before triggering (active-waiting tasks and incomplete
+tasks) then it already belongs to a flow.
+
+Otherwise the triggered tasks will be given all active flow numbers, by
+default. Use the --flow option to trigger a new flow instead, or run a one-off
+task independent of any flow, or to assign it to specified flows.
```
Comment (hjoliver) [inline cylc/flow/scripts/trigger.py:48]: NOTE in 2/2 this wording gets tweak again. Don't bother reviewing it here.

**Comment by hjoliver** [comment]
(Rebased and deconflicted)

**Comment by hjoliver** [comment]
Closing this as it didn't get merged before 2/2 was available, which is built on top of this branch.  I'll mark this as superseded by the other PR.

**Pull request closed without merging**